Add advanced filtering

klauspost opened the pull request:
When running tests where multiple request types are used, it if quite difficult to set up proper filtering.

To help this, I made it possible to chain multiple filters and check URL and method.

Furthermore the URL can now be modified.

For future-proofing the Request and Response structs can now be extended with more stuff without breaking compatibility.

If you like this, I can add some tests for the new functionality.

klauspost commented:
After doing some examples I've decided to re-write some of the stuff...

It is a bit too clunky to use now.

klauspost commented:
@seborama I have added examples for using the specific filtering. If there is more I can do, please let me know.

seborama commented:
I just noticed you have also posted some examples...

klauspost commented:
To give the overall picture, I need some tests to test functionality that can do multiple calls to the same endpoint. 

The endpoint will return unique IDs, so I need to be able to flexibly adjust content of the returned bodies. URLs typically have IDs as well, so I need to be able to still match content even though IDs change.

A huge pain point was that it wasn't possible to do URL/method matching, so transforms would only happen when specific combinations were hit. This will record requests with request transformations applied, so "transformed" requests would be stored and allow for more flexible matching.

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+// Original filter is unmodified.
+func (r RequestFilter) OnPath(pathRegEx string) RequestFilter {
+	if pathRegEx == "" {
+		pathRegEx = "*"
```
This isn't a valid regex.
Use `"."` or `".*"`

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+// Original filter is unmodified.
+func (r ResponseFilter) OnPath(pathRegEx string) ResponseFilter {
+	if pathRegEx == "" {
+		pathRegEx = "*"
```
`"*"` is not a valid regex.
use `"."` or `".*"`

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+	*r = v
+}
+
+// First one or more filters before the current ones.
```
`Prepend` rather?

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+}
+
+// First one or more filters before the current ones.
+func (r ResponseFilters) First(filters ...ResponseFilter) ResponseFilters {
```
`Prepend` rather?

seborama commented on govcr.go:
```diff
@@ -270,8 +532,23 @@ type RequestFilterFunc func(http.Header, []byte) (*http.Header, *[]byte)
 // Return values:
 //  - value 1 - Response's amended header
 //  - value 2 - Response's amended body
+// Deprecated: Use ResponseFilterFunc instead.
```
If we're going to go down this road, let's get rid off `ResponseFilterFunc` altogether.
This would form v3 of govcr.

klauspost replied:
👍 Just didn't want to break it unless you were fine with it.

seborama commented on govcr.go:
```diff
@@ -28,12 +30,21 @@ const defaultCassettePath = "./govcr-fixtures/"
 	Client            *http.Client
 	ExcludeHeaderFunc ExcludeHeaderFunc
+
+	// Deprecated, use RequestFilter
```
If we're going to go down this road, let's get rid off `ResponseFilterFunc` altogether.
This would form v3 of govcr.

seborama commented on govcr.go:
```diff
@@ -28,12 +30,21 @@ const defaultCassettePath = "./govcr-fixtures/"
 
 	// ResponseFilterFunc can be used to modify the header of the response.
 	// This is useful when a fingerprint is exchanged and expected to match between request and response.
+	// Deprecated, use ResponseFilter
```
If we're going to go down this road, let's get rid off `ResponseFilterFunc` altogether.
This would form v3 of govcr.

seborama commented on examples/example6.go:
```diff
@@ -0,0 +1,112 @@
+	}
+
+	// print outcome.
+	fmt.Println("Status code:", resp.StatusCode, " (should be 404 on real and 202 on replay)")
```
This should form the basis of a formal go example test (as in the root folder `xxx_examplen_test.go` and some unit tests too.

seborama commented on examples/example6.go:
```diff
@@ -0,0 +1,112 @@
+
+		// Add header if method was "GET"
+		govcr.ResponseFilter(func(resp govcr.Response) govcr.Response {
+			resp.Header.Add("method-was-get", "true")
```
best to use a different value such as `present` or `added` etc
when using 'true', this is converted to a boolean by the http client which then cannot be displayed by Println().

klauspost replied:
I am not sure what you mean, it prints fine:

```
method-was-post: true (should be true on replay on POST)
```

seborama replied:
Ha, odd! On my environment the word 'true' is missing (go 1.11.1)

seborama commented on examples/example6.go:
```diff
@@ -0,0 +1,112 @@
+
+		// Add header if method was "POST"
+		govcr.ResponseFilter(func(resp govcr.Response) govcr.Response {
+			resp.Header.Add("method-was-post", "true")
```
best to use a different value such as `present` or `added` etc
when using 'true', this is converted to a boolean by the http client which then cannot be displayed by Println().

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+// ResponseFilters is a slice of ResponseFilter
+type ResponseFilters []ResponseFilter
+
+// ResponseContext provides the response parameters.
```
Not a ResponseContext

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+	}
+}
+
+// AddHeaderValue will add or overwrite a header to the request
```
Method says RequestAddHeaderValue?

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+	}
+}
+
+// DeleteHeaderKeys will delete one or more header keys on the request
```
Method says RequestDeleteHeaderKeys

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+	}
+}
+
+// ResponseDeleteHeader will delete one or more headers on the response when returned from vcr playback.
```
Method says ResponseDeleteHeaderKeys

seborama commented on govcr.go:
```diff
@@ -255,8 +266,259 @@ type ExcludeHeaderFunc func(key string) bool
+	}
+}
+
+// Chain one or more filters after the current one and return as single filter.
```
Invalid comment: method name is Append

klauspost commented:
>the units tests have not been written - ahem...

Indeed. Didn't want to spend the time upfront if you didn't like the concept.

>new example tests (as in go xxx_example.go in the root of the project) are needed and some of the visual checks in particular in examples/example6.go can be migrated there.

👍 

>the doc needs updating to reflect the removal of FilterFunc and introduction of Filters
👍 

>I'd probably move the RequestFilter and ResponseFilter receiver functions each in their own separate go source file.

👍 

> It shows a deep transfiguration of the recorded track at replay time.

Technically it is not much that couldn't be achieved before, just in a more available interface.

Things like testing external interfaces often includes complex logic, and having it available for advanced needs can be a positive.

seborama commented:
Sounds good. I realise there's quite a bit in my comments. Don't feel coerced into doing all. I can pick up some of it if you like.

klauspost commented:
>Sounds good. I realise there's quite a bit in my comments. Don't feel coerced into doing all. I can pick up some of it if you like.

No problem. I have added tests for all new functions and added the examples as tests as well.

Made a few tweaks. I removed "Append" and changed "Prepend" to not return the updated slice - seems to just be a way of forgetting to re-assign the returned value.

I can't really tell if `ExcludeHeaderFunc` is the same as `RequestDeleteHeaderKeys`. As far as I can tell it is. If so, I guess it can just be removed.

seborama commented:
Thanks. Let me know when you're happy with the PR and I'll merge it. It would be nice if at least the docs were updated :smile:

seborama commented:
Sorry I'm reading this on my mobile and I seem to consistently fail to read your messages in whole. I just noticed you said you're done. I'll merge when I get a chance.

Thanks for your time and for sharing your contributions. I'm sure many people will be grateful too.

Bonne chance!
